### PR TITLE
Introduce TracePos type for trace positions

### LINKED ISSUE #272: Introduce new type for trace positions
Many internal structures refer to trace positions; this is an important type of values.  In the code, it is sometimes `u32`, sometimes `usize` (!).  While suffixes of such variables usually have "_pos" suffix, having a dedicated type will improve code with type guarantees.

Other numbers like stream positions/generation might deserve their types too.

### PR DESCRIPTION
**Pull request by monoid**
Many internal structures refer to trace positions; this is an important type of values. In the code, it is sometimes u32, sometimes usize (!). While suffixes of such variables usually have "_pos" suffix, having a dedicated type does improve code with type guarantees.

Closes #272

### CODE REVIEW

File: air/tests/test_module/features/errors/last_error.rs [hunk 1]
```diff
@@ -211,7 +211,7 @@ fn variable_names_shown_in_error() {
     let trace = trace_from_result(&result);
 
     assert_eq!(
-        trace[1],
+        trace.as_ref()[1],
```
Comment (mikevoronov) [inline air/tests/test_module/features/errors/last_error.rs:214]: You implemented the `Index` trait for `ExecutionTrace`, why not just `trace[1]`?

File: crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs [hunk 2]
```diff
@@ -62,9 +64,10 @@ impl TraceSlider {
         Some(result)
     }
 
-    pub(crate) fn set_position_and_len(&mut self, position: usize, subtrace_len: usize) -> KeeperResult<()> {
+    pub(crate) fn set_position_and_len(&mut self, position: TracePos, subtrace_len: usize) -> KeeperResult<()> {
+        let pos: usize = position.into();
         // it's possible to set empty subtrace_len and inconsistent position
-        if subtrace_len != 0 && position + subtrace_len > self.trace.len() {
+        if subtrace_len != 0 && pos + subtrace_len > self.trace.len() {
```
Comment (mikevoronov) [inline crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs:70]: Let's implement `cheched_add` for `TracePos` instead.

File: crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs [hunk 3]
```diff
@@ -80,7 +83,7 @@ impl TraceSlider {
     }
 
     pub(crate) fn set_subtrace_len(&mut self, subtrace_len: usize) -> KeeperResult<()> {
-        let trace_remainder = self.trace.len() - self.position;
+        let trace_remainder = Into::<TracePos>::into(self.trace.len()) - self.position;
```
Comment (mikevoronov) [inline crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs:86]: I think that it'd be more readable to convert position to `usize`.

File: air/tests/test_module/features/errors/last_error.rs (hunk 1 above)
Comment (monoid) [reply air/tests/test_module/features/errors/last_error.rs]: `ExecutionTrace` is to be indexable by `TracePos` only.

File: crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs (hunk 2 above)
Comment (monoid) [reply crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs:70]: Would you like to .unwrap() it or bubble as error?

File: air/tests/test_module/features/errors/last_error.rs (hunk 1 above)
Comment (monoid) [reply air/tests/test_module/features/errors/last_error.rs]: Yep, I've changed on other places, but overlooked this one.

Comment (monoid) [reply air/tests/test_module/features/errors/last_error.rs]: Yep, it has been changed :)

File: crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs (hunk 2 above)
Comment (monoid) [reply crates/air-lib/trace-handler/src/data_keeper/trace_slider.rs:70]: Will be resolved in the #275.

**Review by mikevoronov — APPROVED** [review]
utACK

**Pull request merged**